### ISSUE TITLE
Logbook forgeability

### ISSUE THREAD

**Issue by burdiyan** [qri-io/qri#1126]
I find Qri’s logbook very similar to other peer-to-peer systems that are based on immutable, append-only logs, like [Scuttlebutt](https://scuttlebutt.nz), [Textile](https://github.com/textileio/go-threads), [DAT](https://dat.foundation/) and etc.

But, in all these systems logs are truly append-only and unforgeable. Each message is signed by the owners private key and each message references the hash of the previous message (not exactly in all of the mentioned systems, but the idea is the same).

Although looks like in Qri it’s not the same. I understand that Qri logs are hierarchical, and are signed given a full opset. So adding an operation would produce a different signature. Is the property of immutability and verifiable authorship something beyond the scope of Qri, or am I missing something?

**Comment by b5** [comment]
Fantastic question! Thanks so much for posting

>  Is the property of immutability and verifiable authorship something beyond the scope of Qri, or am I missing something?

Qri does have immutable, verifiable authorship, it's one level lower in the stack: attached to each _dataset commit_. Commits include the user's peerID & a base64-encoded signature of the commit timestamp & body hash, and a reference to the previous commit hash. All of these fields are set when creating a new version.

Logbook is built to solve a different problem. It's meant to be a lightweight coordination layer for tracking dataset naming, and discovering versions without moving around dataset data.

You're exactly right that logbook signatures describe the entire opset, and adding an operation changes the signature. This design choice takes more inspiration from TLS than from append-only-logs, because logbooks are optimized to be updated & sent frequently. Keeping the signature of the entire log provides both a cache invalidation heuristic, keeps the logs themselves much lighter and forces peers echoing logs to retain a single signature from the original author if they'd like to reprovide logbook data.

Having both systems together gives us nice characteristics. Logbooks assist in fast dataset discovery, and have built-in mechanisms for elevating logbook data above gossip, while the full dataset data gives properties of PKI-backed attribution.

Does that make sense? I'd be happy to go into more detail!

**Comment by burdiyan** [comment]
@b5 Thanks for your detailed response! All that does make sense, and is interesting!

But I think I'm still missing something. Looking at the files in my dataset, I don't see any links to previous versions of the dataset. File `commit.json` does indeed have signature and author's peer ID, but no link to the previous commit.

The only place where I was able to find link to the previous version is actually the logbook. But from what you say, I understand that commits should also refer to the previous versions, right?

What I don't understand basically is how "full dataset data gives properties of PKI-backed attribution", without the actual logbook? Or is that there's no way to get a dataset without the logbook (I'm maybe thinking about getting somehow raw IPFS blocks, and not being aware where to get the logbook from, is that even possible)?

**Comment by b5** [comment]
Yep! This is a quirk of the `commit` component. The `previous` field is on the _datatset_ object, not in the commit. We're intentionally saving the `previous` field in the commit component for when we work out how commits with multiple parents will work. For now if you run

```
qri get me/$DATASET_NAME | grep previousPath
```

the hash of the previous commit will show up (if it exists. Initial commits have no `previousPath`).

**Comment by burdiyan** [comment]
@b5 Oh, I see, that's right. Previous path indeed is in dataset.json. But there's no signature.

The thing is that, I'm working on a project where attribution and authorship are quite important things. So I was wondering if using qri's oplog approach could be more interesting than a typical append-only log with signature chaining for each message (like in Scuttlebutt).
